<?xml version="1.0"?>
<robot>
    <!--Add gazebo specific definitions here-->

    <gazebo reference="base_arm_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>

    <gazebo reference="link_1_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>

    <gazebo reference="link_2_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>

    <gazebo reference="link_3_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>

    <gazebo reference="right_finger_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>

    <gazebo reference="left_finger_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>


    <gazebo reference="wheel_right_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>500000.0</kp>
        <kd>10.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>0.1</maxVel>
        <fdir1>1 0 0</fdir1>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>

    <gazebo reference="wheel_left_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>500000.0</kp>
        <kd>10.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>0.1</maxVel>
        <fdir1>1 0 0</fdir1>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>

    <gazebo reference="casta_wheel_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>1000000.0</kp>
        <kd>100.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>1.0</maxVel>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>

    <gazebo>
        <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
            <robot_sim_type>gazebo_ros2_control/GazeboSystem</robot_sim_type>
            <parameters>$(find robot_cleaner_sim)/config/controller_config.yaml</parameters>
        </plugin>
    </gazebo>
    
    <ros2_control name="GazeboSystem" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>

        <joint name="base_arm_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="link_1_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="link_2_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="link_3_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>


        <joint name="left_finger_joint">
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="right_finger_joint">
            <param name="mimic">left_finger_joint</param>
            <param name="multiplier">1</param>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
    </ros2_control>

    <gazebo>
        <plugin name="mobile_base_controller" filename="libgazebo_ros_diff_drive.so">
            <!-- <ros>
                <namespace>/demo</namespace>

                <remapping>cmd_vel:=cmd_demo</remapping>
                <remapping>odom:=odom_demo</remapping>

            </ros> -->
            <update_rate>100</update_rate>

            <!--1 for differential drive; 2 for skid steer drive-->

            <num_wheel_pairs>1</num_wheel_pairs>
            <left_joint>wheel_left_joint</left_joint>
            <right_joint>wheel_right_joint</right_joint>
            <wheel_separation>0.35</wheel_separation>
            <wheel_diameter>0.04</wheel_diameter>
            <max_wheel_torque>50</max_wheel_torque>
            <max_wheel_acceleration>1.0</max_wheel_acceleration>

            <publish_odom>true</publish_odom>
            <publish_odom_tf>true</publish_odom_tf>
            <publish_wheel_tf>false</publish_wheel_tf>

            <odometry_frame>odom</odometry_frame>
            <robot_base_frame>mobile_base_link</robot_base_frame>
        </plugin>
    </gazebo>
</robot>

